CRUD operations

PaleRoom opened the pull request.

Commit:
add services, add db connection

GUR-ok commented on src/main/java/ru/ncs/DemoShop/controllers/ProductController.java:
```diff
@@ -0,0 +1,138 @@
+        return ResponseEntity.ok(HttpStatus.OK);
+    }
+
+    @PostMapping("/{id}/delete")
```
- [x] @DeleteMapping. путь поменять на просто {id}

GUR-ok commented on src/main/java/ru/ncs/DemoShop/repositories/ProductRepository.java:
```diff
@@ -0,0 +1,17 @@
+public interface ProductRepository extends JpaRepository<Product, UUID> {
+    Optional<Product> findByName(String name);
+
+    Optional<Product> findById(UUID id);
```
есть по дефолту

GUR-ok commented on src/main/java/ru/ncs/DemoShop/DTO/ProductDTO.java:
```diff
@@ -0,0 +1,76 @@
+import java.time.LocalDateTime;
+import java.util.UUID;
+
+public class ProductDTO {
```
подключить lombok, поставить аннотацию Data или Value

GUR-ok commented on src/main/java/ru/ncs/DemoShop/DTO/ProductDTO.java:
```diff
@@ -0,0 +1,76 @@
+
+public class ProductDTO {
+
+   private UUID id;
```
форматирование поехало. есть полезные команды в идее ctrl+alt+o и ctrl+alt+l это автовыравнивание и автоудаление неиспользуемых импортов

GUR-ok commented on src/main/java/ru/ncs/DemoShop/DTO/ProductDTO.java:
```diff
@@ -0,0 +1,76 @@
+    @NotEmpty(message = "product name should be not Empty")
+    private String name;
+
+
```
пропуски по 1 строке а не 2

GUR-ok commented on src/main/java/ru/ncs/DemoShop/DTO/ProductDTO.java:
```diff
@@ -0,0 +1,76 @@
+    private String category;
+
+
+    @Min(value = 0, message = "Price should be at least 0 or higher")
```
@Positive

GUR-ok commented on src/main/java/ru/ncs/DemoShop/DTO/ProductDTO.java:
```diff
@@ -0,0 +1,76 @@
+    @Min(value = 0, message = "Amount should be at least 0 or higher")
+    private int amount;
+
+    public String getName() {
```
убрать сеттеры и геттеры после подключения ломбока. при подключении ломбока нужно сделать enable annotation processing!

GUR-ok commented on src/main/java/ru/ncs/DemoShop/DemoShopApplication.java:
```diff
@@ -0,0 +1,18 @@
+		SpringApplication.run(DemoShopApplication.class, args);
+	}
+	@Bean
+	public ModelMapper modelMapper() {
```
вынести в конфиг. тут не должно быть ничего лишнего

GUR-ok commented on src/main/java/ru/ncs/DemoShop/controllers/ProductController.java:
```diff
@@ -0,0 +1,138 @@
+import java.util.stream.Collectors;
+
+@RestController
+@RequestMapping("/products")
```
добавить ломбоковскую @RequiredArgsConstructor

GUR-ok commented on src/main/java/ru/ncs/DemoShop/controllers/ProductController.java:
```diff
@@ -0,0 +1,138 @@
+    private final ModelMapper modelMapper;
+
+    @Autowired
+    public ProductController(ProductService productService,
```
убрать конструктор после ломбока

GUR-ok commented on src/main/java/ru/ncs/DemoShop/controllers/ProductController.java:
```diff
@@ -0,0 +1,138 @@
+        return convertToProductDTO(productService.findOne(id));
+    }
+
+    @PostMapping("/create")
```
просто @Postmapping

GUR-ok replied:
@ResponseStatus(value = HttpStatus.CREATED)
будет 201 возвращаться. поправить тесты постман если надо

GUR-ok commented on src/main/java/ru/ncs/DemoShop/controllers/ProductController.java:
```diff
@@ -0,0 +1,138 @@
+    }
+
+    @PostMapping("/create")
+    public ResponseEntity<?> create(@RequestBody @Valid ProductDTO productDTO, BindingResult bindingResult) {
```
- [x] убрать responseEntity везде. 
- [x] Для контроллера тоже сделать интерфейс.
- [x]  В интерфейсе поставить сваггеровские аннотации сколько сможешь)
- [x] проверить работу валидации полей для классов реквестов. Что если поле не должно быть null, не должна проходить валидация входящего запроса. Использовать javax validation аннотации, типа @notnull, notblank для стрингов. Попробовать в постамене передать null для например name в запросе на создание

GUR-ok replied:
тут будет UUID просто возвращаться

GUR-ok commented on src/main/java/ru/ncs/DemoShop/controllers/ProductController.java:
```diff
@@ -0,0 +1,138 @@
+        //return ResponseEntity.ok(HttpStatus.OK);
+    }
+
+    @PostMapping("/{id}/update")
```
@PutMapping или патч. путь поменять на просто "/{id}"

GUR-ok commented on src/main/java/ru/ncs/DemoShop/controllers/ProductController.java:
```diff
@@ -0,0 +1,138 @@
+    }
+
+
+    @ExceptionHandler
```
вынести все exceptionHandler в отдельный класс и пометить как @ControllerAdvice. Пока засунуть в пакет exception

GUR-ok commented on src/main/java/ru/ncs/DemoShop/controllers/ProductController.java:
```diff
@@ -0,0 +1,138 @@
+
+
+    @ExceptionHandler
+    private ResponseEntity<ProductErrorResponse> handleException(ProductNotCreatedException e) {
```
ResponseEntity в exceptionHandler остаются

GUR-ok commented on src/main/java/ru/ncs/DemoShop/controllers/ProductController.java:
```diff
@@ -0,0 +1,138 @@
+package ru.ncs.DemoShop.controllers;
```
пакеты переименовать на единственное число. не controllers а controller

GUR-ok commented on src/main/java/ru/ncs/DemoShop/models/Product.java:
```diff
@@ -0,0 +1,113 @@
+    @Column(name = "id", columnDefinition = "BINARY(16)")
+    @GeneratedValue
+
+
```
форматирование

GUR-ok commented on src/main/java/ru/ncs/DemoShop/models/Product.java:
```diff
@@ -0,0 +1,113 @@
+@Table(name = "products")
+public class Product {
+    @Id
+    @Column(name = "id", columnDefinition = "BINARY(16)")
```
@Column(name = "id", columnDefinition = "BINARY(16)", updatable = false, nullable = false)

GUR-ok commented on src/main/java/ru/ncs/DemoShop/models/Product.java:
```diff
@@ -0,0 +1,113 @@
+
+    private UUID id;
+
+    @Column(name = "product_name")
```
nullable = false

GUR-ok commented on src/main/java/ru/ncs/DemoShop/models/Product.java:
```diff
@@ -0,0 +1,113 @@
+    @Column(name = "product_description")
+    private String description;
+
+    @Column(name = "product_category")
```
nullable = false

GUR-ok commented on src/main/java/ru/ncs/DemoShop/models/Product.java:
```diff
@@ -0,0 +1,113 @@
+    private String description;
+
+    @Column(name = "product_category")
+    private String category;
```
давай не стринг а енум. придумай какие-нибудь категории.

GUR-ok commented on src/main/java/ru/ncs/DemoShop/models/Product.java:
```diff
@@ -0,0 +1,113 @@
+    private String category;
+
+    @Column (name = "product_price")
+    @Min(value = 0, message = "Price should be at least 0 or higher")
```
@Positive

GUR-ok commented on src/main/java/ru/ncs/DemoShop/models/Product.java:
```diff
@@ -0,0 +1,113 @@
+    @Column(name = "product_category")
+    private String category;
+
+    @Column (name = "product_price")
```
nullable = false

GUR-ok commented on src/main/java/ru/ncs/DemoShop/models/Product.java:
```diff
@@ -0,0 +1,113 @@
+    @Min(value = 0, message = "Price should be at least 0 or higher")
+    private double price;
+
+    @Column (name = "product_amount")
```
nullable = false

GUR-ok replied:
nullable = false

GUR-ok commented on src/main/java/ru/ncs/DemoShop/models/Product.java:
```diff
@@ -0,0 +1,113 @@
+    @Min(value = 0, message = "Amount should be at least 0 or higher")
+    private int amount;
+
+    @Column (name="amount_update")
```
nullable = false

GUR-ok commented on src/main/java/ru/ncs/DemoShop/models/Product.java:
```diff
@@ -0,0 +1,113 @@
+        this.amount = amount;
+    }
+
+    public void setId(UUID id) {
```
добавить ломбоковские @Getter @Setter @ToString @noargsconstructor, @allargsconstructor. Конструкторы убрать из кода

GUR-ok commented on src/main/java/ru/ncs/DemoShop/repositories/ProductRepository.java:
```diff
@@ -0,0 +1,17 @@
+
+    Optional<Product> findById(UUID id);
+
+    void deleteById(UUID id);
```
есть по дефолту

GUR-ok commented on src/main/java/ru/ncs/DemoShop/services/ProductErrorResponse.java:
```diff
@@ -0,0 +1,28 @@
+        this.timestamp = timestamp;
+    }
+
+    public String getMessage() {
```
ломбок @Value

GUR-ok commented on src/main/java/ru/ncs/DemoShop/services/ProductService.java:
```diff
@@ -0,0 +1,74 @@
+@Service
+@Transactional(readOnly = true)
+
+public class ProductService {
```
- [x] здесь должно быть ProductServiceImpl implements ProductService. создать интерфейс ProductService

GUR-ok commented on src/main/java/ru/ncs/DemoShop/services/ProductService.java:
```diff
@@ -0,0 +1,74 @@
+        return productRepository.findAll();
+    }
+
+    public Product findOne(UUID id) {
```
Сервис пусть возвращает Dto

GUR-ok commented on src/main/java/ru/ncs/DemoShop/services/ProductService.java:
```diff
@@ -0,0 +1,74 @@
+        return foundProduct.orElse(null);
+    }
+//
+//    public Product findOne(String name) {
```
убрать закоменченые строки

GUR-ok commented on src/main/java/ru/ncs/DemoShop/services/ProductService.java:
```diff
@@ -0,0 +1,74 @@
+    public void delete(UUID id) {
+        productRepository.deleteById(id);
+    }
+
```
форматирование должно быть красивое)

GUR-ok commented on src/main/java/ru/ncs/DemoShop/util/ProductNotCreatedException.java:
```diff
@@ -0,0 +1,5 @@
+package ru.ncs.DemoShop.util;
+
+public class ProductNotCreatedException extends RuntimeException{
```
- [x] вынести эксепшены в отдельный пакет exception

GUR-ok commented on src/main/java/ru/ncs/DemoShop/util/ProductValidator.java:
```diff
@@ -0,0 +1,36 @@
+    private final ProductService productService;
+
+    @Autowired
+    public ProductValidator(ProductService productService) {
```
lombok @RequiredArgsConstructor

GUR-ok commented on src/main/resources/applicatiom.properties.EDIT_THIS:
```diff
@@ -0,0 +1,12 @@
+
```
переделать на yaml. ьудет называться application.yaml

GUR-ok commented on src/main/resources/applicatiom.properties.EDIT_THIS:
```diff
@@ -0,0 +1,12 @@
+
+server.port=7070
+spring.datasource.driverClassName=org.postgresql.Driver
+spring.datasource.url=*your_database_URL_like_jdbc:postgresql://localhost:5432/Shop_db*
```
что означают звездочки?

GUR-ok commented on src/main/resources/db/changelog/create-changeset-product-table.xml:
```diff
@@ -0,0 +1,37 @@
+            <column name="product_description" type="varchar">
+            </column>
+
+            <column name="product_category" type="varchar(64)">
```
nullable="false"

GUR-ok commented on src/main/resources/db/changelog/create-changeset-product-table.xml:
```diff
@@ -0,0 +1,37 @@
+            <column name="product_category" type="varchar(64)">
+            </column>
+
+            <column name="product_price" type="double">
```
nullable="false"

GUR-ok commented on src/main/resources/db/changelog/create-changeset-product-table.xml:
```diff
@@ -0,0 +1,37 @@
+            <column name="product_price" type="double">
+            </column>
+
+            <column name="product_amount" type="bigint">
```
nullable="false"

GUR-ok commented on src/main/resources/db/changelog/create-changeset-product-table.xml:
```diff
@@ -0,0 +1,37 @@
+            <column name="product_amount" type="bigint">
+            </column>
+
+            <column name="amount_update" type="datetime">
```
nullable="false"

GUR-ok commented on src/main/resources/db/changelog/create-changeset-product-table.xml:
```diff
@@ -0,0 +1,37 @@
+
+            </column>
+
+
```
- [x] нужно добавить еще поле цена

GUR-ok commented on src/main/resources/db/changelog/insert-changeset-product-table.xml:
```diff
@@ -0,0 +1,22 @@
+    <property name="now" value="now()" dbms="postgresql"/>
+
+    <changeSet id="insert-into-product" author="Nicolas">
+
```
не обязательно добавлять какую-то запись, но для тестирования ок

GUR-ok commented on pom.xml:
```diff
@@ -0,0 +1,71 @@
+			<scope>test</scope>
+
+		</dependency>
+
```
добавить open api (swagger)

GUR-ok commented on src/main/java/ru/ncs/DemoShop/controllers/ProductController.java:
```diff
@@ -0,0 +1,138 @@
+package ru.ncs.DemoShop.controllers;
+
```
- [x] в пакете controller сделать два пакета request и response.  контроллер должен возвращать и принимать объекты, классы которых создать в этих пакетах. для create и update свои классы. типа CreateProductRequest. В пакете response может ничего не быть, тогда он не нужен. Пусть контроллер возвращает productDto и List ProductDto. Либо сделать GetProductsResponse где будет одно поле List ProductDto на случай запроса всех товаров. Это удобно, если мы захотим возвращать только айдишники. Переделаем GetProductsResponse и все. Можно, например добавить туда поле size
- [x] в пакете сервис тоже сделать два пакета data и immutable. immutable для реквестов, data для возвращаемых значений. в контроллере CreateProductRequest конвертить в ImmutableCreateProductRequest. ProductDto переместить в пакет data в пакете service. Сервис пусть возвращает productDto или список productDto для findAll. 
- [x] конвертацию в контроллере делать при помощи spring ConversionService https://www.baeldung.com/spring-type-conversions. Конвертация в сервисе пусть останется через modelmapper

GUR-ok commented on src/main/java/ru/ncs/DemoShop/controllers/ProductController.java:
```diff
@@ -0,0 +1,138 @@
+        this.modelMapper = modelMapper;
+    }
+
+    @GetMapping("/show")
```
просто getmapping. не забыть пофиксить постман тесты

GUR-ok commented on src/main/java/ru/ncs/DemoShop/services/ProductErrorResponse.java:
```diff
@@ -0,0 +1,28 @@
+package ru.ncs.DemoShop.services;
```
вынести в новый пакет с exception

GUR-ok commented on src/main/java/ru/ncs/DemoShop/services/ProductService.java:
```diff
@@ -0,0 +1,74 @@
+        return foundProduct.orElseThrow(ProductNotFoundException::new);
+    }
+
+    public Product findOneForValidator(UUID id) {
```
я бы этот метод убрал, а в валидаторе обработал через try-catch, вызывая предыдущий метод

GUR-ok commented on src/main/java/ru/ncs/DemoShop/services/ProductService.java:
```diff
@@ -0,0 +1,74 @@
+    }
+
+    @Transactional
+    public void update(Product product) {
```
здесь не void а UUID

GUR-ok commented on src/main/java/ru/ncs/DemoShop/services/ProductService.java:
```diff
@@ -0,0 +1,74 @@
+        if (productRepository.findById(product.getId()).get().getAmount() != product.getAmount())
+            product.setAmountUpdatedAt(LocalDateTime.now());
+        productRepository.save(product);
+    }
```
return product.getId()

GUR-ok commented on src/main/resources/applicatiom.properties.EDIT_THIS:
```diff
@@ -0,0 +1,12 @@
+spring.jpa.properties.hibernate.show_sql=true
+spring.liquibase.enabled = true
+spring.liquibase.change-log=classpath:/db/changelog/changelog.xml
+spring.liquibase.clear-checksums=true
```
почитай про  spring.jpa.hibernate.ddl-auto: validate . Всегда ставят validate, для отладки можно create или create-drop или update

GUR-ok reviewed (commented):
- [x] Давай сделаем, чтобы контроллер возвращал дто с датой в формате 2022/Ноябрь/14 23:12:45.898 😉

Commit:
Add service layers

GUR-ok commented on pom.xml:
```diff
@@ -55,6 +55,29 @@
+
+		<!-- https://mvnrepository.com/artifact/io.springfox/springfox-swagger-ui -->
+		<dependency>
+			<groupId>io.springfox</groupId>
```
Попробуй без этой зависимости

GUR-ok commented on src/main/java/ru/ncs/DemoShop/controller/ProductControllerImpl.java:
```diff
@@ -0,0 +1,57 @@
+@RestController
+@RequestMapping("/products")
+@RequiredArgsConstructor
+
```
Убрать пустую строку

GUR-ok commented on src/main/java/ru/ncs/DemoShop/controller/request/CreateProductRequest.java:
```diff
@@ -0,0 +1,30 @@
+
+    @Positive(message = "Amount should be at least 0 or higher")
+    private int amount;
+
```
Убрать пустые строки

GUR-ok commented on src/main/java/ru/ncs/DemoShop/controller/response/GetProductResponse.java:
```diff
@@ -0,0 +1,25 @@
+
+@Data
+@Builder
+
```
Пустая строка

GUR-ok commented on src/main/java/ru/ncs/DemoShop/exception/ExceptionHandler.java:
```diff
@@ -0,0 +1,37 @@
+
+
+@ControllerAdvice
+public class ExceptionHandler {
```
Назови GlobalExceptionHandler

GUR-ok commented on src/main/java/ru/ncs/DemoShop/exception/ExceptionHandler.java:
```diff
@@ -0,0 +1,37 @@
+        return new ResponseEntity<>(response, HttpStatus.BAD_REQUEST); // 401
+    }
+
+    @org.springframework.web.bind.annotation.ExceptionHandler
```
Тогда тут будет аннотация нормальная

GUR-ok commented on src/main/java/ru/ncs/DemoShop/interfaces/ControllerConverter.java:
```diff
@@ -0,0 +1,22 @@
+import java.util.List;
+import java.util.UUID;
+
+
```
Переместить в пакет с контроллером

GUR-ok commented on src/main/java/ru/ncs/DemoShop/interfaces/ProductService.java:
```diff
@@ -0,0 +1,26 @@
+
+    @Transactional
+     UUID save(ImmutableCreateProductRequest immutableCreateProductRequest);
+
```
Лишние строки тут. А между 15 и 16 встпвить пустую строку.

GUR-ok commented on src/main/java/ru/ncs/DemoShop/model/Product.java:
```diff
@@ -0,0 +1,48 @@
+@Entity
+@Table(name = "products")
+    @Getter
+        @Setter
```
Не такой елочкой)

GUR-ok commented on src/main/java/ru/ncs/DemoShop/model/ProductCategoryEnum.java:
```diff
@@ -0,0 +1,13 @@
+
+import lombok.NoArgsConstructor;
+
+@NoArgsConstructor
```
Ненужная аннотация тут

GUR-ok commented on src/main/java/ru/ncs/DemoShop/service/DatabaseConverterImpl.java:
```diff
@@ -0,0 +1,50 @@
+import ru.ncs.DemoShop.service.data.ProductDTO;
+import ru.ncs.DemoShop.service.immutable.ImmutableCreateProductRequest;
+import ru.ncs.DemoShop.service.immutable.ImmutableUpdateProductRequest;
+
```
Такие конверторы тоже пойдут. Норм. Но давай чтоб научиться новому, сделаем через спринговый ConvertionService

GUR-ok commented on src/main/java/ru/ncs/DemoShop/service/ProductErrorResponse.java:
```diff
@@ -0,0 +1,13 @@
+import lombok.Value;
+
+@Getter @Setter @AllArgsConstructor @Value
+public class ProductErrorResponse {
```
В пакет с исклбчениями. Аннотации в столбик елочкой.

GUR-ok commented on src/main/java/ru/ncs/DemoShop/service/data/ProductDTO.java:
```diff
@@ -0,0 +1,20 @@
+import java.time.LocalDateTime;
+import java.util.UUID;
+
+@Data
```
@Value вместо data. Все поля должны быть final

GUR-ok commented on src/main/java/ru/ncs/DemoShop/service/immutable/ImmutableCreateProductRequest.java:
```diff
@@ -0,0 +1,16 @@
+import lombok.Data;
+import ru.ncs.DemoShop.model.ProductCategoryEnum;
+
+@Data
```
@Value вместо data. Все поля должны быть final

GUR-ok commented on src/main/java/ru/ncs/DemoShop/service/immutable/ImmutableUpdateProductRequest.java:
```diff
@@ -0,0 +1,18 @@
+
+import java.util.UUID;
+
+@Data
```
@Value вместо data. Все поля должны быть final

GUR-ok commented on src/main/resources/application.yml:
```diff
@@ -0,0 +1,22 @@
+  liquibase:
+    enabled: true
+    change-log: 'classpath:/db/changelog/changelog.xml'
+    clear-checksums: true
```
Не нужно очищать контрольные суммы

GUR-ok commented on src/main/java/ru/ncs/DemoShop/interfaces/ProductController.java:
```diff
@@ -0,0 +1,33 @@
+import javax.validation.Valid;
+import java.util.List;
+import java.util.UUID;
+
```
Переместить в пакет с контроллером

GUR-ok commented on src/main/java/ru/ncs/DemoShop/interfaces/ProductService.java:
```diff
@@ -0,0 +1,26 @@
+@Service
+@Transactional(readOnly = true)
+public interface ProductService {
+
```
Давай переместим в пакет с сервисами

Commit:
1.ConversionService added
2.Exception handlers fixed
3.Validation fixed
4.Clear-checksum:false
5.Interfaces locations fixed
6.Postman json updated
